presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -114,20 +120,46 @@
         implements HiveMetastore
 {
     private final ThriftHiveMetastoreStats stats;
+    private final MBeanExporter exporter;
     private final HiveCluster clientProvider;
     private final Function<Exception, Exception> exceptionMapper;
+    private final Map<HostAndPort, ThriftHiveMetastoreStats> metastoreStats;
+    private final boolean isHmsImpersonationEnabled;
 
     @Inject
-    public ThriftHiveMetastore(HiveCluster hiveCluster)
+    public ThriftHiveMetastore(HiveCluster hiveCluster, MetastoreClientConfig config, MBeanExporter exporter)
     {
-        this(hiveCluster, new ThriftHiveMetastoreStats(), identity());
+        this(hiveCluster, config, exporter, new ThriftHiveMetastoreStats(), identity());
     }
 
-    public ThriftHiveMetastore(HiveCluster hiveCluster, ThriftHiveMetastoreStats stats, Function<Exception, Exception> exceptionMapper)
+    public ThriftHiveMetastore(HiveCluster hiveCluster, MetastoreClientConfig config, MBeanExporter exporter, ThriftHiveMetastoreStats stats, Function<Exception, Exception> exceptionMapper)
     {
+        this.isHmsImpersonationEnabled = config.isMetastoreImpersonationEnabled();
+        this.exporter = requireNonNull(exporter, "exporter is null");
         this.clientProvider = requireNonNull(hiveCluster, "hiveCluster is null");
         this.stats = requireNonNull(stats, "stats is null");
         this.exceptionMapper = requireNonNull(exceptionMapper, "exceptionMapper is null");
+        ImmutableMap.Builder<HostAndPort, ThriftHiveMetastoreStats> metastoreStatsBuilder = ImmutableMap.builder();
+        for (HostAndPort hostAndPort : hiveCluster.getAddresses()) {
+            metastoreStatsBuilder.put(hostAndPort, new ThriftHiveMetastoreStats());
+        }
+        this.metastoreStats = metastoreStatsBuilder.build();
+        exportMetastoreStatistics();
+    }
+
+    private void exportMetastoreStatistics()
+    {
+        for (HostAndPort key : metastoreStats.keySet()) {
+            String name = ObjectNames.builder(ThriftHiveMetastore.class)
+                    .withProperty("host", key.getHost())
```

Comment:
Connector id should be included when exporting stats to avoid clashes when more than a sigle Hive catalog is deployed. See https://github.com/prestodb/presto/blob/master/presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/ThriftMetastoreModule.java#L71